scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -108,24 +109,346 @@ object ExecutionOptimizationRules {
+      // type
+      def toExecution(ex: ExH): Execution[H]
+      def fn: H => A
+
+      // convert back to the original Execution type
+      final def execution: Execution[A] =
+        this match {
+          case Single(a) => a
+          case notSingle =>
+            Mapped(toExecution(executions), fn)
+        }
+
+      def zip[B](that: FlattenedZip[Execution, B]): FlattenedZip[Execution, (A, B)]
+      def map[B](fn: A => B): FlattenedZip[Ex, B] =
+        this match {
+          case MapZip(f1, fn1) =>
+            MapZip(f1, ComposedFunctions.ComposedMapFn(fn1, fn))
+          case notMap =>
+            MapZip(notMap, fn)
+        }
+
+      // How many writes are there in this FlattenedZip
+      def writeCount: Int = {
+        def isW[A](ex: Execution[A]): Int =
+          ex match {
+            case WriteExecution(_, _, _) => 1
+            case _ => 0
+          }
+        @annotation.tailrec
+        def loop(fz: FlattenedZip[Execution, Any], acc: Int): Int =
+          fz match {
+            case Single(w) => acc + isW(w)
+            case Many(h, rest, _) => loop(rest, acc + isW(h))
+            case MapZip(of, _) => loop(of, acc)
+          }
+
+        loop(this, 0)
+      }
+
+      /**
+       * The type of this ensures that the head of the HList
+       * must be a WriteExecution. Additionally, none of
+       * the other items are, but we don't prove that in the types,
+       * instead we prove that with tests
+       */
+      def optimize: Option[FlattenedZip[WriteExecution, A]]
+    }
+
+    object FlattenedZip {
+      import HList._
+
+      /**
+       * Here is the simplest FlattenZip: just a single Execution
+       */
+      final case class Single[+Ex[x] <: Execution[x], A](ex: Ex[A]) extends FlattenedZip[Ex, A] {
+        type H = A :: HNil.type
+        type ExH = Execution[A] :: HNil.type
+        def executions = ::(ex, HNil)
+        def toExecution(ex: Execution[A] :: HNil.type): Execution[A :: HNil.type] =
+          ex.head.map(ConsFn(HNil))
+
+        def fn = HeadFn()
+        def zip[B](that: FlattenedZip[Execution, B]) =
+          that match {
+            case s@Single(ex2) => Many[Execution, A, B](ex, s)
+            case m@Many(_, _, _) =>
+              def go[X, Y](m: Many[Execution, X, Y, B]): FlattenedZip[Execution, (A, B)] = {
+                val sub0: SubTypes[(X, Y), B] = m.sub
+                val flat0: FlattenedZip[Execution, (A, (X, Y))] =
+                  Many[Execution, A, (X, Y)](ex, Single(m.head).zip(m.rest))
+                type F[+X] = FlattenedZip[Execution, (A, X)]
+                sub0.liftCo[F](flat0)
+              }
+              go(m)
+            case MapZip(left, fn) =>
+              zip(left).map(ZipMap.MapRight(fn))
+          }
+        def optimize: Option[FlattenedZip[WriteExecution, A]] =
+          ex match {
+            case w@WriteExecution(_, _, _) => Some(Single(w))
+            case _ => None
```

Comment:
For this my understanding is that you can never hit this optimize from the external call site so it returning none isn't lossy? (it looks like for many if it sees a None on a single it would keep it around as the rest).
If you didn't guard the outer with:
```
          wc = flat.writeCount
          // only optimize if there is more the 1 write, otherwise we create an infinite loop
          if (wc > 1)
```
this could be an issue? (just trying to make sure i understand whats happening)

Reply:
I'll add a comment, but we call optimize recursively, and some of those branches can hit None. Also note the type it returns: `Option[FlattenedZip[WriteExecution, A]]` so, you have no way of making `FlattenedZip[WriteExecution, A]` if there is not at least one `WriteExecution` in the zip, since that type requires the first execution be a write.